layer/input_buffer.cc

```diff
@@ -0,0 +1,209 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "input_buffer.h"
+
+#include <cassert>
+#include <cstdio>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "logging.h"
+
+#if defined(__unix__)
+#include <fcntl.h>
+#include <sys/io.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#endif
+
+namespace performancelayers {
+
+namespace {
+// Cross-platfrom input buffer implementation that reads file contents using
+// fopen/fread/fclose.
+class FileInputBufferImpl : public InputBuffer::InputBufferImplBase {
+ public:
+  static absl::StatusOr<std::unique_ptr<InputBuffer::InputBufferImplBase>>
+  Create(const std::string& path) {
+    struct FileCloser {
+      void operator()(FILE* file) {
+        if (file) fclose(file);
+      }
+    };
+    std::unique_ptr<FILE, FileCloser> handle(fopen(path.c_str(), "rb"));
+    if (!handle) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to fopen file for read: ", path));
+    }
+
+    fseek(handle.get(), 0, SEEK_END);
+    const auto file_size = ftell(handle.get());
+    fseek(handle.get(), 0, SEEK_SET);
+    if (file_size < 0) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to read file size: ", path));
+    }
+
+    const auto buffer_size = static_cast<size_t>(file_size);
+    FileInputBufferImpl res;
+    res.buffer_.resize(buffer_size);
+    if (fread(res.buffer_.data(), 1, buffer_size, handle.get()) !=
+        buffer_size) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to read file buffer contents: ", path));
+    }
+
+    return std::make_unique<FileInputBufferImpl>(std::move(res));
+  }
+
+  absl::Span<const uint8_t> GetBuffer() const override { return buffer_; }
+  ~FileInputBufferImpl() override = default;
+
+ private:
+  std::vector<uint8_t> buffer_;
+};
+
+#if defined(__unix__)
+// Unix memory mapped input buffer implementation.
+class UnixMemMappedInputBufferImpl : public InputBuffer::InputBufferImplBase {
+ public:
+  static absl::StatusOr<std::unique_ptr<InputBuffer::InputBufferImplBase>>
+  Create(const std::string& path) {
+    UnixMemMappedInputBufferImpl res;
+    res.file_descriptor_ = open(path.c_str(), O_RDONLY);
+    if (res.file_descriptor_ == -1) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to open file for read: ", path));
+    }
+
+    struct stat s = {};
+    if (fstat(res.file_descriptor_, &s) != 0) {
+      return absl::UnavailableError(
+          absl::StrCat("Failed to stat file: ", path));
+    }
+    size_t size = s.st_size;
+    if (size == 0) {
+      // We cannot mmap empty files, but we can return an empty buffer without
+      // mmaping the underlying file.
+      return std::make_unique<UnixMemMappedInputBufferImpl>(std::move(res));
+    }
+
+    void* data = mmap(0, size, PROT_READ, MAP_PRIVATE, res.file_descriptor_, 0);
```

Comment:
I am confused .. if we use `PROT_READ`, we have to only read something from it. As far as I remember, writing sth on it causes a fault.

If I understand it correctly, why don't we use `const void* data` instead of `void* data`?
Likewise, I am not sure we can use it for `uint8_t* buffer_start_ = nullptr;` (not `const`).

Reply:
Yes, this is read only. The only reason why I didn't use a const pointer is because `munmap` expects a `void*` parameter, and I didn't want to introduce `const_casts` for little benefit. This class only ever exposes the memory through a const span.

What do you think?

Reply:
If we use it only insider this class, it would be fine. Could you please add a comment to explain this is a read only buffer?

Reply:
Done. Thanks for the suggestion.